README.md
# [Tooltip](https://github.com/fengyuanchen/tooltip)

> A simple jQuery tooltip plugin.

- [Demo](http://fengyuanchen.github.io/tooltip)


## Main

```
dist/
├── tooltip.css     (2 KB)
├── tooltip.min.css (2 KB)
├── tooltip.js      (9 KB)
└── tooltip.min.js  (4 KB)
```



## Getting started

### Quick start

Two quick start options are available:

- [Download the latest release](https://github.com/fengyuanchen/tooltip/archive/master.zip).
- Clone the repository: `git clone https://github.com/fengyuanchen/tooltip.git`.



### Installation

Include files:

```html
<script src="/path/to/jquery.js"></script><!-- jQuery is required -->
<link  href="/path/to/tooltip.css" rel="stylesheet">
<script src="/path/to/tooltip.js"></script>
```


### Usage

Initialize with `$.fn.tooltip` method.

```js
// With option
$('body').tooltip({
  autoshow: true,
  content: 'Hello, world!'
});

// With method
$('body').tooltip('show', 'Hello, world!');

// Shortcut
$('body').tooltip('Hello, world!');
```


## Options

```js
// Set tooltip options
$().tooltip(options);

// Change the global default options
$.fn.tooltip.setDefaults(options);
```


### autohide

- Type: `Boolean`
- Default: `true`

Hide the tooltip automatically when timeout. requires `duration` option to be set.


### autoshow

- Type: `Boolean`
- Default: `true`

Show the tooltip automatically when initialize. requires `content` option to be set.


### content

- Type: `String`
- Default: `''`

Tooltip content


### countdown

- Type: `Boolean`
- Default: `false`

Show countdown view. requires `duration` option to be set.


### dismissible

- Type: `Boolean`
- Default: `false`

Allow to dismiss the tooltip by click an attached close button before it is closed automatically.


### duration

- Type: `Number`
- Default: `3000`

Define the time of the tooltip showing (3 seconds by default).


### fixed

- Type: `Boolean`
- Default: `true`

Fix the tooltip (Only available for `body` element).


### offset

- Type: `Number`
- Default: `10`

Offset of the tooltip from its parent.


### position

- Type: `String`
- Default: `'center top'`
- Options:
    `'left top'`,    `'center top'`,    `'right top'`,
    `'left middle'`, `'center middle'`, `'right middle'`,
    `'left bottom'`, `'center bottom'`, `'right bottom'`

Position of the tooltip (horizontal and vertical)


### style

- Type: `String`
- Default: `'default'`
- Options: `'default'`, `'primary'`, `'success'`, `'info'`, `'warning'`, `'danger'`

Style of the tooltip


### zIndex

- Type: `Number`
- Default: `1024`

Z-index of the tooltip.



## Methods


```js
$().tooltip('method', argument1, , argument2, ..., argumentN)
```


### show([content[, options]])

- **content** (optional):
  - Type: `String`
  - Tooltip's [content](#content)
- **options** (optional):
  - Type: `String` or `Object`
  - Tooltip's [style](#style) or tooltip's [options](#options)

Show the tooltip with new content.

```js
$().tooltip('show', 'Hello, world!');
$().tooltip('show', 'Hello, world!', 'primary');
$().tooltip('show', 'Hello, world!', {
  style: 'primary',
  zIndex: 2048
});
```


**Shortcut:** `$().tooltip(content, options)`

- Only for `show` method
- The first argument must be a string (not one of the methods)

```js
$().tooltip('Hello, world!', 'primary');
```


### hide()

Hide the tooltip.


### destroy()

Destroy the tooltip.



## Events


### show.tooltip

This event fires immediately when the `show` instance method is called.


### shown.tooltip

This event is fired when the tooltip has been made visible to the user (will wait for CSS transitions to complete).


### hide.tooltip

This event is fired immediately when the hide instance method has been called.


### hidden.tooltip

This event is fired when the tooltip has finished being hidden from the user (will wait for CSS transitions to complete).



## No conflict

If you have to use other plugin with the same namespace, just call the `$.fn.tooltip.noConflict` method to revert to it.

```html
<script src="other-plugin.js"></script>
<script src="tooltip.js"></script>
<script>
  $.fn.tooltip.noConflict();
  // Code that uses other plugin's "$().tooltip" can follow here.
</script>
```



## Browser Support

- Chrome (latest 2)
- Firefox (latest 2)
- Internet Explorer 8+
- Opera (latest 2)
- Safari (latest 2)

As a jQuery plugin, you also need to see the [jQuery Browser Support](http://jquery.com/browser-support/).



## [License](LICENSE.md)

Released under the [MIT](http://opensource.org/licenses/mit-license.html) license.


CHANGELOG.md
# Changelog


## 0.0.2 (Jun 22, 2015)

- Support a shortcut for showing message content.
- Improve the `options` argument of "show" method.


## 0.0.1 (Jun 10, 2015)

- Supports eleven options: "autohide", "autoshow", "content", "countdown", "dismissible", "duration", "fixed", "offset", "position", "style", "zIndex"
- Supports three methods: "show", "hide", "destroy"
- Supports four events: "show.tooltip", "shown.tooltip", "hide.tooltip", "hidden.tooltip"


gulpfile.js
'use strict';

var gulp = require('gulp'),
    plugins = require('gulp-load-plugins')(),
    pkg = require('./package'),
    scripts = {
      all: [
        'gulpfile.js',
        'src/*.js',
        'docs/js/*.js'
      ],
      src: 'src/*.js',
      docs: 'docs/js',
      dest: 'dist'
    },
    styles = {
      all: [
        'src/*.css',
        'docs/css/*.css'
      ],
      src: 'src/*.css',
      docs: 'docs/css',
      dest: 'dist'
    },
    replacement = {
      regexp: /@\w+/g,
      filter: function (placeholder) {
        switch (placeholder) {
          case '@VERSION':
            placeholder = pkg.version;
            break;

          case '@YEAR':
            placeholder = (new Date()).getFullYear();
            break;

          case '@DATE':
            placeholder = (new Date()).toISOString();
            break;
        }

        return placeholder;
      }
    };

gulp.task('jshint', function () {
  return gulp.src(scripts.all)
  .pipe(plugins.jshint('src/.jshintrc'))
  .pipe(plugins.jshint.reporter('default'));
});

gulp.task('jscs', function () {
  return gulp.src(scripts.all)
  .pipe(plugins.jscs('src/.jscsrc'));
});

gulp.task('js', ['jshint', 'jscs'], function () {
  return gulp.src(scripts.src)
  .pipe(plugins.replace(replacement.regexp, replacement.filter))
  .pipe(gulp.dest(scripts.docs))
  .pipe(gulp.dest(scripts.dest))
  .pipe(plugins.rename('tooltip.min.js'))
  .pipe(plugins.uglify({
    preserveComments: 'some'
  }))
  .pipe(gulp.dest(scripts.dest));
});

gulp.task('jscopy', function () {
  return gulp.src(scripts.src)
  .pipe(gulp.dest(scripts.dest));
});

gulp.task('csslint', function () {
  return gulp.src(styles.all)
  .pipe(plugins.csslint('src/.csslintrc'))
  .pipe(plugins.csslint.reporter());
});

gulp.task('css', ['csslint'], function () {
  return gulp.src(styles.src)
  .pipe(plugins.replace(replacement.regexp, replacement.filter))
  .pipe(plugins.autoprefixer())
  .pipe(plugins.csscomb())
  .pipe(gulp.dest(styles.docs))
  .pipe(gulp.dest(styles.dest))
  .pipe(plugins.rename('tooltip.min.css'))
  .pipe(plugins.minifyCss())
  .pipe(gulp.dest(styles.dest));
});

gulp.task('csscopy', function () {
  return gulp.src(styles.src)
  .pipe(gulp.dest(styles.dest));
});

gulp.task('docs', function () {
  return gulp.src('docs/**')
  .pipe(gulp.dest('_gh_pages'));
});

gulp.task('release', ['js', 'css'], function () {
  return gulp.src('dist/*')
  .pipe(gulp.dest('_releases/' + pkg.version));
});

gulp.task('watch', function () {
  gulp.watch(scripts.src, ['jscopy']);
  gulp.watch(styles.src, ['csscopy']);
});

gulp.task('default', ['watch']);


js/tooltip.js
/*!
 * Tooltip v0.0.2
 * https://github.com/fengyuanchen/tooltip
 *
 * Copyright (c) 2015 Fengyuan Chen
 * Released under the MIT license
 *
 * Date: 2015-06-22T03:35:32.356Z
 */

(function (factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as anonymous module.
    define('tooltip', ['jquery'], factory);
  } else if (typeof exports === 'object') {
    // Node / CommonJS
    factory(require('jquery'));
  } else {
    // Browser globals.
    factory(jQuery);
  }
})(function ($) {

  'use strict';

  var $window = $(window),

      // Constants
      NAMESPACE = 'tooltip',

      // Classes
      CLASS_FIXED = NAMESPACE + '-fixed',
      CLASS_RELATIVE = NAMESPACE + '-relative',
      CLASS_ACTIVE = NAMESPACE + '-active',
      CLASS_DISMISSIBLE = NAMESPACE + '-dismissible',
      CLASS_COUNTDOWN = NAMESPACE + '-countdown',

      // Events
      EVENT_CLICK = 'click.' + NAMESPACE,
      EVENT_RESIZE = 'resize.' + NAMESPACE,
      EVENT_SHOW = 'show.' + NAMESPACE,
      EVENT_SHOWN = 'shown.' + NAMESPACE,
      EVENT_HIDE = 'hide.' + NAMESPACE,
      EVENT_HIDDEN = 'hidden.' + NAMESPACE;

  function isString(s) {
    return typeof s === 'string';
  }

  function Tooltip(element, options) {
    this.$element = $(element);
    this.options = $.extend({}, Tooltip.DEFAULTS, $.isPlainObject(options) && options);
    this.active = false;
    this.animating = false;
    this.dismissing = false;
    this.classes = false;
    this.center = false;
    this.middle = false;
    this.isBody = false;
    this.isStatic = false;
    this.init();
  }

  Tooltip.prototype = {
    constructor: Tooltip,

    init: function () {
      var options = this.options,
          $this = this.$element;

      this.initialOptions = $.extend({}, this.options);
      this.isBody = $this.is('body');
      this.isStatic = !this.isBody && $this.css('position') === 'static';
      this.$tooltip = $(Tooltip.TEMPLATE.dialog);

      $this.on(EVENT_CLICK, '[data-dismiss="' + NAMESPACE + '"]', $.proxy(this.dismiss, this));
      $window.on(EVENT_RESIZE, $.proxy(this.resize, this));

      if (options.autoshow) {
        this.show(options.content);
      }
    },

    // Example Usage:
    //   - show(content, options)
    //   - show(content, style)
    //   - show(options)
    show: function (content, options) {
      var $tooltip = this.$tooltip,
          showEvent = $.Event(EVENT_SHOW),
          position,
          classes;

      this.$element.trigger(showEvent);

      if (showEvent.isDefaultPrevented()) {
        return;
      }

      if (options) {
        if (isString(options) && $.inArray(options, Tooltip.STYLES) > -1) {
          options = {
            style: options
          };
        }
      } else {
        if ($.isPlainObject(content)) {
          options = content;
          content = '';
        }
      }

      this.options = options = $.extend({}, this.initialOptions, $.isPlainObject(options) && options);
      content = content || options.content;

      // Type conversion
      options.duration = Number(options.duration) || 3000;
      options.offset = Number(options.offset) || 10;
      options.zIndex = Number(options.zIndex) || 1024;

      if (!isString(content) || !content.length) {
        return;
      }

      position = isString(options.position) ? options.position.split(/\W+/) : ['center', 'top'];
      this.center = $.inArray('center', position) > -1;
      this.middle = $.inArray('middle', position) > -1;

      classes = $.map(position, function (n) {
        return NAMESPACE + '-' + n;
      });

      if (options.fixed && this.isBody) {
        classes.unshift(CLASS_FIXED);
      }

      classes.push($.inArray(options.style, Tooltip.STYLES) > -1 ? (NAMESPACE + '-' + options.style) : options.style);

      if (options.countdown) {
        content += Tooltip.TEMPLATE.countdown;
      }

      if (options.dismissible) {
        classes.push(CLASS_DISMISSIBLE);
        content += Tooltip.TEMPLATE.button;
      }

      $tooltip
      .removeClass(this.classes + ' ' + CLASS_ACTIVE)
      .addClass(this.classes = classes.join(' '))
      .html(content)
      .appendTo(this.$element.toggleClass(CLASS_RELATIVE, this.isStatic));

      if (options.countdown) {
        this.countdown();
      }

      this.active = true;
      this.resize();
      $tooltip.addClass(CLASS_ACTIVE);


      if (this.animating) {
        clearTimeout(this.animating);
      }

      this.animating = setTimeout($.proxy(this.shown, this), 150); // 150ms for fade in
    },

    shown: function () {
      var options = this.options;

      this.animating = false;

      if (options.autohide) {
        if (this.dismissing) {
          clearTimeout(this.dismissing);
        }

        this.dismissing = setTimeout($.proxy(this.hide, this), options.duration);
      }

      this.$element.trigger(EVENT_SHOWN);
    },

    hide: function () {
      var hideEvent = $.Event(EVENT_HIDE);

      this.$element.trigger(hideEvent);

      if (hideEvent.isDefaultPrevented()) {
        return;
      }

      this.active = false;
      this.$tooltip.removeClass(CLASS_ACTIVE);

      if (this.animating) {
        clearTimeout(this.animating);
      }

      this.animating = setTimeout($.proxy(this.hidden, this), 150); // 150ms for fade out
    },

    hidden: function () {
      this.animating = false;
      this.$tooltip.empty().detach();
      this.$element.removeClass(CLASS_RELATIVE).trigger(EVENT_HIDDEN);
    },

    resize: function () {
      var options = this.options,
          $tooltip = this.$tooltip,
          offset = options.offset;

      if (this.active) {
        $tooltip.removeAttr('style').css({
          marginTop: this.middle ? -$tooltip.outerHeight() / 2 : offset,
          marginRight: offset,
          marginBottom: offset,
          marginLeft: this.center ? (offset - $tooltip.outerWidth() / 2) : offset,
          zIndex: options.zIndex
        });
      }
    },

    dismiss: function () {
      if (this.dismissing) {
        clearTimeout(this.dismissing);
      }

      if (this.counting) {
        clearTimeout(this.counting);
      }

      this.dismissing = false;
      this.counting = false;
      this.hide();
    },

    countdown: function () {
      var $countdown = this.$tooltip.find('.' + CLASS_COUNTDOWN),
          seconds = this.options.duration / 1000;

      (function countdown() {
        $countdown.text(' (' + seconds + ') ');

        if (seconds-- > 0) {
          setTimeout(countdown, 1000);
        }
      })();
    },

    destroy: function () {
      this.$element.off(EVENT_CLICK, this.dismiss).removeData(NAMESPACE).removeClass(CLASS_RELATIVE);
      $window.off(EVENT_RESIZE, this.resize);
    }
  };

  Tooltip.DEFAULTS = {
    // Auto hide when timeout
    // Type: Boolean
    autohide: true,

    // Auto show after initialized
    // Type: Boolean
    autoshow: true,

    // Tooltip content
    // Type: String
    content: '',

    // Show countdown view
    // Type: Boolean
    countdown: false,

    // Allow to dismiss the tooltip before it is closed automatically
    // Type: Boolean
    dismissible: false,

    // Define the time of the tooltip showing (3 seconds by default)
    // Type: Number
    duration: 3000,

    // Fix the tooltip (Only available for body element)
    // Type: Boolean
    fixed: true,

    // Offset of the tooltip from its parent
    // Type: Number
    offset: 10,

    // Position of the tooltip (horizontal and vertical)
    // Type: String
    // Options:
    //   'left top',    'center top',    'right top',
    //   'left middle', 'center middle', 'right middle',
    //   'left bottom', 'center bottom', 'right bottom'
    position: 'center top',

    // Style of the tooltip
    // Type: String
    // Options: 'default', 'primary', 'success', 'info', 'warning', 'danger'
    style: 'default',

    // Z index of the tooltip
    // Type: Number
    zIndex: 1024
  };

  Tooltip.setDefaults = function (options) {
    $.extend(Tooltip.DEFAULTS, options);
  };

  Tooltip.STYLES = ['default', 'primary', 'success', 'info', 'warning', 'danger'];

  Tooltip.TEMPLATE = {
    dialog: '<div class="tooltip-dialog"></div>',
    countdown: '<span class="tooltip-countdown"></span>',
    button: '<button type="button" class="tooltip-dismiss" data-dismiss="tooltip" aria-label="Close"><span aria-hidden="true">&times;</span></button>'
  };

  // Save the other tooltip
  Tooltip.other = $.fn.tooltip;

  // Register as jQuery plugin
  $.fn.tooltip = function (options) {
    var args = [].slice.call(arguments, 1);

    return this.each(function () {
      var $this = $(this),
          data = $this.data(NAMESPACE),
          fn;

      if (!data) {
        if (/destroy|hide/.test(options)) {
          return;
        }

        $this.data(NAMESPACE, (data = new Tooltip(this, options)));
      }

      if (isString(options)) {
        fn = data[options];

        if (!$.isFunction(fn)) {
          fn = data.show; // Support the shortcut of "show" method
          args.unshift(options);
        }

        fn.apply(data, args);
      }
    });
  };

  $.fn.tooltip.Constructor = Tooltip;
  $.fn.tooltip.setDefaults = Tooltip.setDefaults;

  // No conflict
  $.fn.tooltip.noConflict = function () {
    $.fn.tooltip = Tooltip.other;
    return this;
  };

});
